diff --git a/build-commons-dev.make b/build-commons-dev.make
index dd8068b..581a227 100644
--- a/build-commons-dev.make
+++ b/build-commons-dev.make
@@ -8,5 +8,6 @@ includes[] = "drupal-org-core.make"
 projects[commons][type] = "profile"
 projects[commons][download][type] = "git"
 projects[commons][download][url] = "http://git.drupal.org/project/commons.git"
-projects[commons][patch][] = "http://drupal.org/files/1908812-drupal-org-dev-make.patch"
+; http://drupal.org/node/1908812#comment-7045196
+projects[commons][patch][] = "http://drupal.org/files/1908812-drupal-org-dev-make-17.patch"
 projects[commons][download][branch] = "7.x-3.x"
\ No newline at end of file
diff --git a/build_distro.sh b/build_distro.sh
index f463e87..714f5ca 100755
--- a/build_distro.sh
+++ b/build_distro.sh
@@ -1,54 +1,117 @@
 #!/bin/bash
+repos=(commons_activity_streams commons_featured commons_notices commons_profile_social commons_user_profile_pages commons_body commons_follow commons_notify commons_q_a commons_utility_links commons_bw commons_groups commons_pages commons_radioactivity commons_wikis commons_content_moderation commons_like commons_polls commons_search commons_wysiwyg commons_documents commons_location commons_posts commons_site_homepage commons_events commons_misc commons_profile_base commons_topics commons_social_sharing)
+
 pull_git() {
     cd $BUILD_PATH/commons_profile
+    if [[ -n $RESET ]]; then
+      git reset --hard HEAD
+    fi
     git pull origin 7.x-3.x
 
-    repos=(commons_activity_streams commons_featured commons_notices commons_profile_social commons_user_profile_pages commons_body commons_follow commons_notify commons_q_a commons_utility_links commons_bw commons_groups commons_pages commons_radioactivity commons_wikis commons_content_moderation commons_like commons_polls commons_search commons_wysiwyg commons_documents commons_location commons_posts commons_site_homepage commons_events commons_misc commons_profile_base commons_topics)
-
-    cd $BUILD_PATH/repos
+    cd $BUILD_PATH/repos/modules
     for i in "${repos[@]}"; do
       echo $i
       cd $i
+      if [[ -n $RESET ]]; then
+        git reset --hard HEAD
+      fi
       git pull origin 7.x-3.x
       cd ..
     done
+    cd $BUILD_PATH/repos/themes/commons_origins
+    git pull origin 7.x-3.x
+}
+
+release_notes() {
+  rm -rf rn.txt
+  OUTPUT="<h2>Release Notes for $RELEASE</h2>"
+  cd $BUILD_PATH/commons_profile
+  OUTPUT="$OUTPUT <h3>Commons Profile:</h3> `drush rn --date $FROM_DATE $TO_DATE`"
+
+  cd $BUILD_PATH/repos/modules
+  for i in "${repos[@]}"; do
+    echo $i
+    cd $i
+    RN=`drush rn --date $FROM_DATE $TO_DATE`
+    if [[ -n $RN ]]; then
+      OUTPUT="$OUTPUT <h3>$i:</h3> $RN"
+    fi
+    cd ..
+  done
+  echo $OUTPUT >> $BUILD_PATH/rn.txt
 }
 
 build_distro() {
-    cd $BUILD_PATH
-    rm -rf ./publish
-    # do we have the profile?
-    if [[ -d $BUILD_PATH/commons_profile ]]; then
-      if [[ -d $BUILD_PATH/repos ]]; then
-        drush make commons_profile/build-commons.make --no-cache --working-copy --prepare-install ./publish
-        rm -rf publish/profiles/commons/modules/contrib/commons*
-      else
-        drush make commons_profile/build-commons-dev.make --no-cache --working-copy --prepare-install ./publish
-        mkdir $BUILD_PATH/repos
-        mv publish/profiles/commons/modules/contrib/commons* repos/
-      fi
-      ln -sf $BUILD_PATH/repos/commons* publish/profiles/commons/modules/contrib/
-      chmod -R 777 publish/sites/default
-      # symlink the profile to our dev copy
-      rm -f publish/profiles/commons/*.*
-      rm -rf publish/profiles/commons/images
-      ln -s $BUILD_PATH/commons_profile/* publish/profiles/commons/
-      pull_git
+    if [[ -d $BUILD_PATH ]]; then
+        cd $BUILD_PATH
+        rm -rf ./publish
+        # do we have the profile?
+        if [[ -d $BUILD_PATH/commons_profile ]]; then
+          if [[ -d $BUILD_PATH/repos ]]; then
+            drush make commons_profile/build-commons.make --no-cache --working-copy --prepare-install ./publish
+            rm -rf publish/profiles/commons/modules/contrib/commons*
+            rm -rf publish/profiles/commons/themes/contrib/commons*
+          else
+            drush make commons_profile/build-commons-dev.make --no-cache --working-copy --prepare-install ./publish
+            mkdir $BUILD_PATH/repos
+            mkdir $BUILD_PATH/repos/modules
+            mkdir $BUILD_PATH/repos/themes
+            mv publish/profiles/commons/modules/contrib/commons* repos/modules/
+            mv publish/profiles/commons/themes/contrib/commons* repos/themes/
+          fi
+          ln -sf $BUILD_PATH/repos/modules/commons* publish/profiles/commons/modules/contrib/
+          ln -sf $BUILD_PATH/repos/themes/commons* publish/profiles/commons/themes/contrib/
+          chmod -R 777 publish/sites/default
+          # symlink the profile to our dev copy
+          rm -f publish/profiles/commons/*.*
+          rm -rf publish/profiles/commons/images
+          ln -s $BUILD_PATH/commons_profile/* publish/profiles/commons/
+          pull_git
+        else
+          git clone http://git.drupal.org/project/commons.git commons_profile
+          build_distro $BUILD_PATH
+        fi
     else
-      git clone http://git.drupal.org/project/commons.git commons_profile
+      mkdir $BUILD_PATH
       build_distro $BUILD_PATH
     fi
 }
 
+# This allows users to build the most recent release from git and apply it to their directory. No-dev removes the symlinks and puts in real versions of the module.
+update() {
+    if [[ -d $BUILD_PATH ]]; then
+        cd $BUILD_PATH
+        # do we have the profile?
+        if [[ -d $BUILD_PATH/commons_profile ]]; then
+          if [[ $DEV == dev ]]; then
+            tar -czvf sites-tmp.tar.gz publish/sites/default
+            build_distro $BUILD_PATH
+            tar -zxvf sites-tmp.tar.gz
+          fi
+
+          if [[ $DEV == nodev ]]; then
+            drush make commons_profile/build-commons.make --no-cache --tar publish
+            tar -zxvf publish.tar.gz
+          fi
+        fi
+    else
+      echo "invalid build path"
+      exit 1
+    fi
+}
+
 case $1 in
   pull)
     if [[ -n $2 ]]; then
       BUILD_PATH=$2
+      if [[ -n $3 ]]; then
+       RESET=1
+      fi
     else
       echo "Usage: build_distro.sh pull [build_path]"
       exit 1
     fi
-    pull_git $BUILD_PATH;;
+    pull_git $BUILD_PATH $RESET;;
   build)
     if [[ -n $2 ]]; then
       BUILD_PATH=$2
@@ -57,4 +120,28 @@ case $1 in
       exit 1
     fi
     build_distro;;
-esac
+  update)
+    if [[ -n $2 ]]; then
+      BUILD_PATH=$2
+      if [[ $3 == 'dev' ]] || [[ $3 == 'nodev' ]]; then
+        DEV=$3
+      else
+        DEV='dev'
+      fi
+    else
+      echo "Usage: build_distro.sh update [build_path] [dev|nodev]"
+      exit 1
+    fi
+    update $BUILD_PATH $DEV;;
+  rn)
+    if [[ -n $2 ]] && [[ -n $3 ]] && [[ -n $4 ]] && [[ -n $5 ]]; then
+      BUILD_PATH=$2
+      RELEASE=$3
+      FROM_DATE=$4
+      TO_DATE=$5
+    else
+      echo "Usage: build_distro.sh rn [build_path] [release] [from_date] [to_date]"
+      exit 1
+    fi
+    release_notes $BUILD_PATH $RELEASE $FROM_DATE $TO_DATE;;
+esac
\ No newline at end of file
diff --git a/commons.info b/commons.info
index 50282a8..ab73c36 100644
--- a/commons.info
+++ b/commons.info
@@ -26,6 +26,7 @@ dependencies[] = toolbar
 dependencies[] = ctools
 dependencies[] = custom_search
 dependencies[] = diff
+dependencies[] = email_registration
 dependencies[] = entity
 dependencies[] = entityreference
 dependencies[] = entitycache
@@ -92,4 +93,5 @@ dependencies[] = commons_activity_page
 dependencies[] = commons_wikis_pages
 dependencies[] = commons_misc
 dependencies[] = commons_location
-dependencies[] = commons_wysiwyg
\ No newline at end of file
+dependencies[] = commons_wysiwyg
+dependencies[] = commons_social_sharing
diff --git a/commons.install b/commons.install
index 617cc1d..c0effe6 100644
--- a/commons.install
+++ b/commons.install
@@ -13,9 +13,13 @@
  * @see system_install()
  */
 function commons_install() {
-  // Enable the Skye theme and set it as the default theme.
-  theme_enable(array('adaptivetheme', 'commons_origins'));
+  // Enable the Origins theme and set it as the default.
+  theme_enable(array('adaptivetheme', 'commons_origins', 'adaptivetheme_admin'));
   variable_set('theme_default', 'commons_origins');
+  // Set a default administrative theme:
+  variable_set('admin_theme', 'adaptivetheme_admin');
+  variable_set('node_admin_theme', FALSE);
+  commons_set_default_avatar();
 
   // Place site blocks in the menu_bar and header regions.
   $menu_block = array(
@@ -81,3 +85,38 @@ function commons_install() {
     drupal_form_submit('system_theme_settings', $form_state);
   }
 }
+
+/**
+ * Enable the Origins theme and set it as the default. *
+ */
+function commons_update_3101() {
+  theme_enable(array('adaptivetheme_admin'));
+  variable_set('theme_default', 'commons_origins');
+  variable_set('node_admin_theme', FALSE);
+  return array();
+}
+
+/**
+ * Reset the default avatar if the associated file doesn't exist in the file_managed table.
+ */
+function commons_update_3102() {
+  $reset_avatar = FALSE;
+  $file = variable_get('user_picture_default', '');
+  if(!$file) {
+    $reset_avatar = TRUE;
+  }
+  if(!$reset_avatar) {
+    $results = db_select('file_managed', 'f')
+      ->fields('f', array('fid'))
+      ->condition('filename', $file,'=')
+      ->execute()
+      ->fetchAll();
+    if(count($results) === 0) {
+      $reset_avatar = TRUE;
+    }
+  }
+  if($reset_avatar) {
+    commons_set_default_avatar();
+  }
+  return array();
+}
\ No newline at end of file
diff --git a/commons.profile b/commons.profile
index 45eb6ab..818c45c 100644
--- a/commons.profile
+++ b/commons.profile
@@ -5,6 +5,15 @@
  */
 
 /**
+ * Implements hook_admin_paths_alter().
+ */
+function commons_admin_paths_alter(&$paths) {
+  // Avoid switching between themes when users edit their account.
+  $paths['user'] = FALSE;
+  $paths['user/*'] = FALSE;
+}
+
+/**
  * Implements hook_install_tasks_alter().
  */
 function commons_install_tasks_alter(&$tasks, $install_state) {
@@ -49,6 +58,7 @@ function commons_form_install_configure_form_alter(&$form, $form_state) {
     '#type' => 'fieldset',
     '#title' => st('Acquia'),
     '#description' => st('The !an can supplement the functionality of Commons by providing enhanced site search (faceted search, content recommendations, content biasing, multi-site search, and others using the Apache Solr service), spam protection (using the Mollom service), and more.  A free 30-day trial is available.', array('!an' => l(t('Acquia Network'), 'http://acquia.com/products-services/acquia-network', array('attributes' => array('target' => '_blank'))))),
+    '#weight' => -11,
   );
   $form['server_settings']['enable_acquia_connector'] = array(
     '#type' => 'checkbox',
@@ -101,6 +111,16 @@ function commons_update_projects_alter(&$projects) {
  */
 function commons_install_tasks() {
 
+  //make sure we have more memory than 196M. if not lets try to increase it.
+  if (ini_get('memory_limit') != '-1' && ini_get('memory_limit') <= '196M') {
+    ini_set('memory_limit', '196M');
+  }
+
+  //make sure we have more memory than 196M. if not lets try to increase it.
+  if ((int)ini_get('max_execution_time') != -1 && (int)ini_get('max_execution_time') != 0 && (int)ini_get('max_execution_time') <= 120) {
+    ini_set('max_execution_time', 120);
+  }
+
   $demo_content = variable_get('commons_install_example_content', FALSE);
   $acquia_connector = variable_get('commons_install_acquia_connector', FALSE);
 
@@ -111,7 +131,7 @@ function commons_install_tasks() {
       'run' => $acquia_connector ? INSTALL_TASK_RUN_IF_NOT_COMPLETED : INSTALL_TASK_SKIP,
     ),
     'commons_installer_palette' => array(
-      'display_name' => st('Chose site color palette'),
+      'display_name' => st('Choose site color palette'),
       'display' => TRUE,
       'type' => 'form',
       'function' => 'commons_installer_palette',
@@ -147,7 +167,7 @@ function commons_install_tasks() {
  */
 function commons_installer_palette() {
   $form = array();
-  require_once('profiles/commons/themes/contrib/commons_origins/commons_origins.palettes.inc');
+  require_once(drupal_get_path('theme', 'commons_origins') . '/commons_origins.palettes.inc');
 
   commons_origins_palettes_form($form);
   $form['commons_origins_palette_fieldset']['#collapsible'] = FALSE;
@@ -318,6 +338,12 @@ function commons_anonymous_welcome_text_form_submit($form_id, &$form_state) {
 }
 
 /**
+ * Helper function to generate a machine name similar to the user's full name.
+ */
+function commons_normalize_name($name) {
+  return drupal_strtolower(str_replace(' ','_', $name));
+}
+/**
  * This function generate a demo content
  */
 function commons_demo_content() {
@@ -327,16 +353,18 @@ function commons_demo_content() {
 
   // Create demo Users
   $demo_users = array(
-    'Lou White' => 'Lou White',
-    'George Foreman' => 'George Foreman',
-    'Cesar Ramirez' => 'Cesar Ramirez',
-    'Elinor Dashwood' => 'Elinor Dashwood',
-    'Matt Edmunds' => 'Matt Edmunds',
+    'Jeff Noyes' => 'Jeff Noyes',
+    'Drew Robertson' => 'Drew Robertson',
+    'Lisa Rex' => 'Lisa Rex',
+    'Katelyn Fogarty' => 'Katelyn Fogarty',
+    'Dharmesh Mistry' => 'Dharmesh Mistry',
+    'Erica Ligeski' => 'Erica Ligeski',
   );
 
+
   foreach ($demo_users as $name) {
     list($first_name, $last_name)  = explode(" ", $name);
-    $normalize_name = drupal_strtolower(str_replace(' ','_', $name));
+    $normalize_name = commons_normalize_name($name);
     $password = user_password(8);
 
 
@@ -354,8 +382,10 @@ function commons_demo_content() {
     $fields['field_name_first'][LANGUAGE_NONE][0]['value'] = $first_name;
     $fields['field_name_last'][LANGUAGE_NONE][0]['value'] = $last_name;
 
-
     $demo_users[$name] = user_save('', $fields);
+
+    // Add avatars to demo Users.
+    commons_add_user_avatar($demo_users[$name]);
   }
 
   // Demo Content.
@@ -367,7 +397,7 @@ function commons_demo_content() {
 
   $boston_group->title = 'Boston';
   $boston_group->body[LANGUAGE_NONE][0]['value'] = commons_veggie_ipsum();
-  $boston_group->uid = $demo_users['Lou White']->uid;
+  $boston_group->uid = $demo_users['Jeff Noyes']->uid;
   $boston_group->language = LANGUAGE_NONE;
   $boston_group->created = time() - 604800;
   $boston_group->status = 1;
@@ -380,7 +410,7 @@ function commons_demo_content() {
 
   $nyc_group->title = 'New York City';
   $nyc_group->body[LANGUAGE_NONE][0]['value'] = commons_veggie_ipsum();
-  $nyc_group->uid = $demo_users['Lou White']->uid;
+  $nyc_group->uid = $demo_users['Drew Robertson']->uid;
   $nyc_group->language = LANGUAGE_NONE;
   $nyc_group->status = 1;
   // Make the group 1 week old:
@@ -394,7 +424,7 @@ function commons_demo_content() {
   node_object_prepare($post);
 
   $post->title = 'Best brunch places in Cambridge';
-  $post->uid = $demo_users['George Foreman']->uid;
+  $post->uid = $demo_users['Lisa Rex']->uid;
   $post->language = LANGUAGE_NONE;
   // 1:30 ago.
   $post->created = time() - 5400;
@@ -426,7 +456,7 @@ function commons_demo_content() {
   node_object_prepare($wiki);
   $wiki->created = time() - 604800;
   $wiki->title = 'How to create a veggie burger';
-  $wiki->uid = $demo_users['Matt Edmunds']->uid;
+  $wiki->uid = $demo_users['Dharmesh Mistry']->uid;
   $wiki->language = LANGUAGE_NONE;
   $wiki->body[LANGUAGE_NONE][0]['value'] = "Celtuce quandong gumbo coriander avocado yarrow broccoli rabe parsnip nori mung bean watercress taro pea sprouts cress. Bush tomato water spinach radish green bean okra spinach garlic cress. Cucumber squash tigernut swiss chard celery cabbage beet greens nori groundnut grape melon seakale. Earthnut pea kakadu plum chicory potato plantain fennel gumbo chickweed gourd cauliflower wakame green bean epazote taro quandong. Celery turnip kombu lotus root lettuce sierra leone bologi kale cauliflower gumbo parsnip taro welsh onion melon asparagus green bean beet greens black-eyed pea jícama. Kohlrabi lentil turnip greens plantain bush tomato leek arugula courgette amaranth yarrow.";
   $wiki->body[LANGUAGE_NONE][0]['format'] = filter_default_format();
@@ -453,7 +483,7 @@ function commons_demo_content() {
   node_object_prepare($event);
 
   $event->title = 'Ribfest Boston 2012';
-  $event->uid = $demo_users['Elinor Dashwood']->uid;
+  $event->uid = $demo_users['Katelyn Fogarty']->uid;
   $event->language = LANGUAGE_NONE;
   $event->body[LANGUAGE_NONE][0]['value'] = "<strong>What ignited in 1999 as a community block party has exploded into one of Boston's most anticipated street festivals.</strong> Averaging 50,000 pounds of ribs and BBQ from more than 30 restaurants, Ribfest Boston 2013 is expected to draw more than 50,000 people. As a nationally recognized music festival, we host a hot blend of Indie, pop, Indie Roots, rock and alt country for one of the most unique band lineups in the city. Families can spend the whole weekend in Kids Square to enjoy live entertainment, inflatables, games and more.";
   $event->body[LANGUAGE_NONE][0]['format'] = filter_default_format();
@@ -487,6 +517,43 @@ function commons_demo_content() {
   variable_del('commons_install_example_content');
 }
 
+function commons_add_user_avatar($account) {
+  global $base_url;
+
+  if ($account->uid) {
+    $picture_directory =  file_default_scheme() . '://' . variable_get('user_picture_path', 'pictures');
+    if(file_prepare_directory($picture_directory, FILE_CREATE_DIRECTORY)){
+      $picture_result = drupal_http_request($base_url . '/profiles/commons/images/avatars/avatar-' . commons_normalize_name($account->name) . '.png');
+      $picture_path = file_stream_wrapper_uri_normalize($picture_directory . '/picture-' . $account->uid . '-' . REQUEST_TIME . '.jpg');
+      $picture_file = file_save_data($picture_result->data, $picture_path, FILE_EXISTS_REPLACE);
+
+      // Check to make sure the picture isn't too large for the site settings.
+      $validators = array(
+        'file_validate_is_image' => array(),
+        'file_validate_image_resolution' => array(variable_get('user_picture_dimensions', '85x85')),
+        'file_validate_size' => array(variable_get('user_picture_file_size', '30') * 1024),
+      );
+
+      // attach photo to user's account.
+      $errors = file_validate($picture_file, $validators);
+
+      if (empty($errors)) {
+        // Update the user record.
+        $picture_file->uid = $account->uid;
+        $picture_file = file_save($picture_file);
+        file_usage_add($picture_file, 'user', 'user', $account->uid);
+        db_update('users')
+          ->fields(array(
+          'picture' => $picture_file->fid,
+          ))
+          ->condition('uid', $account->uid)
+          ->execute();
+        $account->picture = $picture_file->fid;
+      }
+    }
+  }
+}
+
 /**
   * Generate some filler content.
  */
@@ -527,6 +594,7 @@ function commons_acquia_connector_enable() {
   $modules = variable_get('commons_install_acquia_modules', array());
   if (!empty($modules)) {
     module_enable($modules, TRUE);
+    commons_clear_messages();
   }
 }
 
@@ -535,10 +603,6 @@ function commons_acquia_connector_enable() {
  */
 function commons_install_finished(&$install_state) {
   // BEGIN copy/paste from install_finished().
-  // Flush all caches to ensure that any full bootstraps during the installer
-  // do not leave stale cached data, and that any content types or other items
-  // registered by the installation profile are registered correctly.
-
   // Remove the bookmarks flag
   $flag = flag_get_flag('bookmarks');
   if($flag) {
@@ -547,8 +611,35 @@ function commons_install_finished(&$install_state) {
     _flag_clear_cache();
   }
 
+  // Flush all caches to ensure that any full bootstraps during the installer
+  // do not leave stale cached data, and that any content types or other items
+  // registered by the installation profile are registered correctly.
   drupal_flush_all_caches();
 
+  // We make custom code for the footer here because we want people to be able to freely edit it if they wish.
+  $footer_body = '<p>'. st('A Commons Community, powered by <a href="@acquia">Acquia</a>', array('@acquia' => url('https://www.acquia.com/products-services/drupal-commons-social-business-software'))) . '</p>';
+
+  $footer_block_text = array(
+    'body' => st($footer_body),
+    'info' => st('Default Footer'),
+    'format' => 'full_html',
+  );
+
+  if (drupal_write_record('block_custom', $footer_block_text)) {
+    $footer_block = array(
+      'module' => 'block',
+      'delta' => $footer_block_text['bid'],
+      'theme' => 'commons_origins',
+      'visibility' => 0,
+      'region' => 'footer',
+      'status' => 1,
+      'pages' => 0,
+      'weight' => 1,
+      'title' => variable_get('site_name', 'Drupal Commons'),
+    );
+    drupal_write_record('block', $footer_block);
+  }
+
   // Remember the profile which was used.
   variable_set('install_profile', drupal_get_profile());
 
@@ -594,3 +685,32 @@ function commons_clear_messages() {
   // Migrate adds its messages under the wrong type, see #1659150.
   drupal_get_messages('ok', TRUE);
 }
+
+/**
+ * Set a default user avatar as a managed file object.
+ */
+function commons_set_default_avatar() {
+  global $base_url;
+  $picture_directory =  file_default_scheme() . '://' . variable_get('user_picture_path', 'pictures');
+  if(file_prepare_directory($picture_directory, FILE_CREATE_DIRECTORY)){
+    $picture_result = drupal_http_request($base_url . '/profiles/commons/images/avatars/user-avatar.png');
+    $picture_path = file_stream_wrapper_uri_normalize($picture_directory . '/picture-default.jpg');
+    $picture_file = file_save_data($picture_result->data, $picture_path, FILE_EXISTS_REPLACE);
+
+    // Check to make sure the picture isn't too large for the site settings.
+    $validators = array(
+      'file_validate_is_image' => array(),
+      'file_validate_image_resolution' => array(variable_get('user_picture_dimensions', '85x85')),
+      'file_validate_size' => array(variable_get('user_picture_file_size', '30') * 1024),
+    );
+
+    // attach photo to user's account.
+    $errors = file_validate($picture_file, $validators);
+
+    if (empty($errors)) {
+      // Update the user record.
+      $picture_file = file_save($picture_file);
+      variable_set('user_picture_default', $picture_path);
+    }
+  }
+}
\ No newline at end of file
diff --git a/drupal-org-core.make b/drupal-org-core.make
index 9eb280b..8f50959 100644
--- a/drupal-org-core.make
+++ b/drupal-org-core.make
@@ -3,7 +3,7 @@ core = 7.x
 
 ; Download Drupal core and apply core patches if needed.
 projects[drupal][type] = "core"
-projects[drupal][version] = "7.19"
+projects[drupal][version] = "7.21"
 
 ; Hide the profiles under /profiles, so Commons is the only one. This allows
 ; the installation to start at the Language selection screen, bypassing a
diff --git a/drupal-org.make b/drupal-org.make
index 1d43dae..4236b85 100644
--- a/drupal-org.make
+++ b/drupal-org.make
@@ -5,32 +5,40 @@ core = 7.x
 
 projects[addressfield][type] = "module"
 projects[addressfield][subdir] = "contrib"
+projects[addressfield][version] = "1.0-beta3"
 
 projects[acquia_connector][type] = "module"
 projects[acquia_connector][subdir] = "contrib"
+projects[acquia_connector][version] = "2.8"
 
 projects[advancedqueue][type] = "module"
 projects[advancedqueue][subdir] = "contrib"
+projects[advancedqueue][version] = "1.0-alpha2"
 
 projects[apachesolr][type] = "module"
 projects[apachesolr][subdir] = "contrib"
+projects[apachesolr][version] = "1.1"
 
-projects[apachesolr_og][version] = "1.x-dev"
 projects[apachesolr_og][type] = "module"
 projects[apachesolr_og][subdir] = "contrib"
+projects[apachesolr_og][download][type] = "git"
+projects[apachesolr_og][download][url] = "http://git.drupal.org/project/apachesolr_og.git"
+projects[apachesolr_og][download][branch] = "7.x-1.x"
+projects[apachesolr_og][download]revision] = "49820b4a4fcff7c1c4efe449da033fb6d8711ac5"
 
 projects[apachesolr_proximity][type] = "module"
 projects[apachesolr_proximity][subdir] = "contrib"
+projects[apachesolr_proximity][version] = "1.0-rc1"
 
-projects[connector][version] = "1.x-dev"
 projects[connector][type] = "module"
 projects[connector][subdir] = "contrib"
+projects[connector][version] = "1.0-beta2"
 
 projects[ckeditor][type] = "module"
 projects[ckeditor][subdir] = "contrib"
 projects[ckeditor][download][type] = "git"
 projects[ckeditor][download][url] = "http://git.drupal.org/project/ckeditor.git"
-; Use Libraries API for ckeditor
+; Use Libraries API for ckeditor.
 ; http://drupal.org/node/1063482#comment-6964504
 projects[ckeditor][download][branch] = "7.x-1.x"
 projects[ckeditor][download][revision] = "f6abbda"
@@ -41,123 +49,177 @@ projects[ckeditor][patch][] = "http://drupal.org/files/ckeditor-install-lib-1898
 
 projects[ctools][type] = "module"
 projects[ctools][subdir] = "contrib"
+projects[ctools][version] = "1.2"
+
 ; http://drupal.org/node/1494860#comment-6204438
 projects[ctools][patch][] = "http://drupal.org/files/ctools-dependent-js-broken-with-jquery-1.7-1494860-30.patch"
 
 projects[custom_search][type] = "module"
 projects[custom_search][subdir] = "contrib"
-projects[custom_search][version] = "1.x-dev"
+projects[custom_search][download][type] = "git"
+projects[custom_search][download][url] = "http://git.drupal.org/project/custom_search.git"
+projects[custom_search][download][branch] = "7.x-1.x"
+projects[custom_search][download][revision] = "e6be0ee6f5f7c7e56f09e79d0d3351fd193c0840"
 
 projects[date][type] = "module"
 projects[date][subdir] = "contrib"
+projects[date][version] = "2.6"
 
 projects[date_facets][type] = "module"
 projects[date_facets][subdir] = "contrib"
+projects[date_facets][version] = "1.0-beta1"
 
+; Keeping this to the latest version, since it should only be used for development.
 projects[devel][version] = "1.x-dev"
 projects[devel][type] = "module"
 projects[devel][subdir] = "contrib"
 
-projects[diff][version] = "2.0"
 projects[diff][type] = "module"
 projects[diff][subdir] = "contrib"
+projects[diff][version] = "3.2"
 
 projects[email_registration][type] = "module"
 projects[email_registration][subdir] = "contrib"
+projects[email_registration][version] = "1.1"
 
 projects[entity][type] = "module"
 projects[entity][subdir] = "contrib"
+projects[entity][version] = "1.0"
+
+; Force LANGUAGE_NONE entities to still display within rendered entities.
+; http://drupal.org/node/1782134
+projects[entity][patch][] = "http://drupal.org/files/entity-translatable_fields_not_overriding_und_with_empty_values.patch"
 
 projects[entitycache][type] = "module"
 projects[entitycache][subdir] = "contrib"
+projects[entitycache][version] = "1.1"
 
-projects[entityreference][version] = "1.x-dev"
 projects[entityreference][type] = "module"
 projects[entityreference][subdir] = "contrib"
+projects[entityreference][download][type] = "git"
+projects[entityreference][download][url] = "http://git.drupal.org/project/entityreference.git"
+projects[entityreference][download][branch] = "7.x-1.x"
+projects[entityreference][download][revision] = "06089a92ee8b269f153064ad3090ce6b464f38aa"
 
+; Profile has no recommended release
 projects[edit_profile][type] = "module"
 projects[edit_profile][subdir] = "contrib"
+projects[edit_profile][version] = "1.0-beta2"
 
 projects[entityreference_prepopulate][type] = "module"
 projects[entityreference_prepopulate][subdir] = "contrib"
+projects[entityreference_prepopulate][version] = "1.2"
 
-projects[features][version] = "2.x-dev"
 projects[features][type] = "module"
 projects[features][subdir] = "contrib"
+projects[features][download][type] = "git"
+projects[features][download][url] = "http://git.drupal.org/project/features.git"
+projects[features][download][branch] = "7.x-2.x"
+projects[features][download][revision] = "63db687c75d7880bb5d2652fa9227e3ca6e4e9f0"
 
 ; Separate fields from field instances.
-; http://drupal.org/node/1064472#comment-6438406
-projects[features][patch][] = "http://drupal.org/files/1064472_features_field_split_23.patch"
+; http://drupal.org/node/1064472#comment-7042482
+projects[features][patch][] = "http://drupal.org/files/features-instances-1064472-59.patch"
 
 projects[flag][type] = "module"
 projects[flag][subdir] = "contrib"
+projects[flag][version] = "2.0"
 
 projects[flag_abuse][type] = "module"
 projects[flag_abuse][subdir] = "contrib"
-projects[flag_abuse][version] = "2.x-dev"
+projects[flag_abuse][version] = "2.0-alpha1"
+
 
 projects[redirect][type] = "module"
 projects[redirect][subdir] = "contrib"
+projects[redirect][version] = "1.0-rc1"
 
-projects[http_client][version] = "2.x-dev"
 projects[http_client][type] = "module"
 projects[http_client][subdir] = "contrib"
+projects[http_client][version] = "2.4"
 
-projects[admin_icons][version] = "1.x-dev"
 projects[admin_icons][type] = "module"
 projects[admin_icons][subdir] = "contrib"
+projects[admin_icons][download][type] = "git"
+projects[admin_icons][download][url] = "http://git.drupal.org/project/admin_icons.git"
+projects[admin_icons][download][branch] = "7.x-1.x"
+projects[admin_icons][download][revision] = "60d9f28801533fecc92216a60d444d89d80e7611"
 
 projects[libraries][type] = "module"
 projects[libraries][subdir] = "contrib"
+projects[libraries][version] = "2.1"
 
 projects[link][type] = "module"
 projects[link][subdir] = "contrib"
+projects[link][version] = "1.1"
 
 projects[menu_attributes][type] = "module"
 projects[menu_attributes][subdir] = "contrib"
+projects[menu_attributes][version] = "1.0-rc2"
 
 projects[message][type] = "module"
 projects[message][subdir] = "contrib"
-projects[message][version] = "1.x-dev"
+projects[message][download][type] = "git"
+projects[message][download][url] = "http://git.drupal.org/project/message.git"
+projects[message][download][branch] = "7.x-1.x"
+projects[message][download][revision] = "5ae12e8c0b7482f8525106825009409f3b9b7140"
 
+; Make message access alterable.
+; http://drupal.org/node/1920560#comment-7080942
+projects[message][patch][] = "http://drupal.org/files/1920560-message-access-alterable.patch"
 
 projects[message_notify][type] = "module"
 projects[message_notify][subdir] = "contrib"
-projects[message_notify][version] = "2.x-dev"
+projects[message_notify][download][type] = "git"
+projects[message_notify][download][url] = "http://git.drupal.org/project/message_notify.git"
+projects[message_notify][download][branch] = "7.x-2.x"
+projects[message_notify][download][revision] = "f8731d574f427dda3bc99e75f541549305655309"
 
 projects[message_subscribe][type] = "module"
 projects[message_subscribe][subdir] = "contrib"
-projects[message_subscribe][version] = "1.x-dev"
+projects[message_subscribe][download][type] = "git"
+projects[message_subscribe][download][url] = "http://git.drupal.org/project/message_subscribe.git"
+projects[message_subscribe][download][branch] = "7.x-1.x"
+projects[message_subscribe][download][revision] = "0063ab0fd922a5dbd2a052114a75a60e10e6ab07"
+
+; Remove "email" as default notifier.
+; http://drupal.org/node/1828184#comment-7081868
+projects[message_subscribe][patch][] = "http://drupal.org/files/email-notifiers-1828184-15.patch"
 
 projects[memcache][type] = "module"
 projects[memcache][subdir] = "contrib"
+projects[memcache][version] = "1.0"
 
 projects[metatag][type] = "module"
 projects[metatag][subdir] = "contrib"
+projects[metatag][version] = "1.0-beta5"
+
 ; Support for rel=author link in head
 ; http://drupal.org/node/1865228#comment-6839604
 projects[metatag][patch][] = "http://drupal.org/files/metatag-n1865228-3.patch"
 
 projects[module_filter][type] = "module"
 projects[module_filter][subdir] = "contrib"
+projects[module_filter][version] = "1.7"
 
 projects[mollom][type] = "module"
 projects[mollom][subdir] = "contrib"
+projects[mollom][version] = "2.4"
 
-projects[oauth][version] = "3.x-dev"
 projects[oauth][type] = "module"
 projects[oauth][subdir] = "contrib"
+projects[oauth][version] = "3.1"
 
-projects[oauthconnector][version] = "1.x-dev"
 projects[oauthconnector][type] = "module"
 projects[oauthconnector][subdir] = "contrib"
+projects[oauthconnector][download][type] = "git"
+projects[oauthconnector][download][url] = "http://git.drupal.org/project/oauthconnector.git"
+projects[oauthconnector][download][branch] = "7.x-1.x"
+projects[oauthconnector][download][revision] = "0ce7ac9614710c0f68d0a58cb4ae4667f8bd6fa7"
 
-projects[og][version] = "2.x-dev"
 projects[og][type] = "module"
 projects[og][subdir] = "contrib"
-; Language tweak for 'user has been added to [group]'
-; http://drupal.org/node/1842830
-projects[og][patch][] = "http://drupal.org/files/og-add-group-message.patch"
+projects[og][version] = "2.1"
 
 ; og_group_ref field should respect og_user_access()
 ; http://drupal.org/node/1902086#comment-7026516
@@ -165,10 +227,12 @@ projects[og][patch][] = "http://drupal.org/files/1902086-og-ref-respect-og-user-
 
 projects[panelizer][type] = "module"
 projects[panelizer][subdir] = "contrib"
-projects[panelizer][version] = "3.0"
+projects[panelizer][version] = "3.1"
 
 projects[panels][type] = "module"
 projects[panels][subdir] = "contrib"
+projects[panels][version] = "3.3"
+
 ; Fatal error: Call to undefined function panels_get_layouts()
 ; http://drupal.org/node/1828684#comment-6694732
 projects[panels][patch][] = "http://drupal.org/files/1828684-layout-fix-6.patch"
@@ -179,21 +243,25 @@ projects[panels][patch][] = "http://drupal.org/files/panels-n1632898-15.patch"
 
 projects[pathauto][type] = "module"
 projects[pathauto][subdir] = "contrib"
+projects[pathauto][version] = "1.2"
 
 projects[pm_existing_pages][type] = "module"
 projects[pm_existing_pages][subdir] = "contrib"
+projects[pm_existing_pages][version] = "1.4"
 
 projects[quicktabs][type] = "module"
 projects[quicktabs][subdir] = "contrib"
-projects[quicktabs][version] = "3.x-dev"
+projects[quicktabs][download][type] = "git"
+projects[quicktabs][download][url] = "http://git.drupal.org/project/quicktabs.git"
+projects[quicktabs][download][branch] = "7.x-3.x"
+projects[quicktabs][download][revision] = "0788eba"
 
 projects[radioactivity][type] = "module"
 projects[radioactivity][subdir] = "contrib"
-projects[radioactivity][version] = "2.x-dev"
-
-; Notice error for accessing an undefined array element
-; http://drupal.org/node/1816252#comment-6617208
-projects[radioactivity][patch][] = "http://drupal.org/files/undefined_array-1816252-1.patch"
+projects[radioactivity][download][type] = "git"
+projects[radioactivity][download][url] = "http://git.drupal.org/project/radioactivity.git"
+projects[radioactivity][download][branch] = "7.x-2.x"
+projects[radioactivity][download][revision] = "aee21dbed4f54d0e626e3c19ecc550bf1ec656f6"
 
 ; Radioactivity not compatible with Memcache module
 ; http://drupal.org/node/1860216
@@ -201,15 +269,19 @@ projects[radioactivity][patch][] = "http://drupal.org/files/radioactivity-memcac
 
 projects[r4032login][type] = "module"
 projects[r4032login][subdir] = "contrib"
+projects[r4032login][version] = "1.5"
 
 projects[rate][type] = "module"
 projects[rate][subdir] = "contrib"
+projects[rate][version] = "1.6"
+
 ; Add widget to node/comment $links
 ; http://drupal.org/node/947516#comment-6979780
 projects[rate][patch][] = "http://drupal.org/files/947516-rate-node-links-15.patch"
 
 projects[realname][type] = "module"
 projects[realname][subdir] = "contrib"
+projects[realname][version] = "1.1"
 
 projects[registration][subdir] = "contrib"
 projects[registration][type] = "module"
@@ -224,187 +296,195 @@ projects[registration][patch][] = "http://drupal.org/files/entity_cache_support-
 
 projects[rules][type] = "module"
 projects[rules][subdir] = "contrib"
+projects[rules][version] = "2.3"
 
 projects[search_facetapi][type] = "module"
 projects[search_facetapi][subdir] = "contrib"
+projects[search_facetapi][version] = "1.0-beta2"
 
 projects[sharethis][type] = "module"
 projects[sharethis][subdir] = "contrib"
+projects[sharethis][version] = "2.5"
 
 projects[facetapi][type] = "module"
 projects[facetapi][subdir] = "contrib"
+projects[facetapi][version] = "1.2"
 
 projects[rich_snippets][type] = "module"
 projects[rich_snippets][subdir] = "contrib"
+projects[rich_snippets][version] = "1.0-beta2"
+
+; Remove snippets from non-node type searches:
+; http://drupal.org/node/1923904#comment-7094488
+projects[rich_snippets][patch][] = "http://drupal.org/files/1923904-search-nodes-only.patch"
 
 projects[schemaorg][type] = "module"
 projects[schemaorg][subdir] = "contrib"
+projects[schemaorg][version] = "1.0-beta3"
 
-projects[strongarm][version] = "2.x-dev"
 projects[strongarm][type] = "module"
 projects[strongarm][subdir] = "contrib"
+projects[strongarm][download][type] = "git"
+projects[strongarm][download][url] = "http://git.drupal.org/project/strongarm.git"
+projects[strongarm][download][branch] = "7.x-2.x"
+projects[strongarm][download][revision] = "5a2326ba67e59923ecce63d9bb5e0ed6548abdf8"
 
 projects[timeago][type] = "module"
 projects[timeago][subdir] = "contrib"
 projects[timeago][version] = "2.x-dev"
+projects[timeago][download][type] = "git"
+projects[timeago][download][url] = "http://git.drupal.org/project/timeago.git"
+projects[timeago][download][branch] = "7.x-2.x"
+projects[timeago][download][revision] = "768ea66"
 
 ; Provide a dedicated date type:
 ; http://drupal.org/node/1427226#comment-6638836
 projects[timeago][patch][] = "http://drupal.org/files/1427226-timeago-date-type.patch"
 
+; Redundant call to drupal_add_js() when Libraries module is enabled.
+; http://drupal.org/node/1945942
+projects[timeago][patch][] = "http://drupal.org/files/timeago-libraries-a.patch"
+
 projects[title][type] = "module"
 projects[title][subdir] = "contrib"
+projects[title][version] = "1.0-alpha7"
 
 projects[token][type] = "module"
 projects[token][subdir] = "contrib"
+projects[token][version] = "1.5"
+
 
 projects[views][type] = "module"
 projects[views][subdir] = "contrib"
-projects[views][version] = "3.x-dev"
+projects[views][version] = "3.6"
+
 
 projects[views_field_view][type] = "module"
 projects[views_field_view][subdir] = "contrib"
+projects[views_field_view][version] = "1.0"
 
 projects[views_load_more][type] = "module"
 projects[views_load_more][subdir] = "contrib"
+projects[views_load_more][version] = "1.1"
 
 projects[views_bulk_operations][type] = "module"
 projects[views_bulk_operations][subdir] = "contrib"
+projects[views_bulk_operations][version] = "3.1"
 
 projects[views_litepager][type] = "module"
 projects[views_litepager][subdir] = "contrib"
+projects[views_litepager][version] = "3.0"
 
+; We have the version of voting api at the top so it doesn't get included in our dev make patch.
+projects[votingapi][version] = "2.11"
 projects[votingapi][type] = "module"
 projects[votingapi][subdir] = "contrib"
 
 ; Commons contrib modules:
+; We don't tag a release here because want them to auto-increment.. and we are the ones who test it.
 
+; We don't have a release for this?!?!
 projects[commons_activity_streams][type] = "module"
 projects[commons_activity_streams][subdir] = "contrib"
-projects[commons_activity_streams][version] = "3.x-dev"
 
 projects[commons_body][type] = "module"
 projects[commons_body][subdir] = "contrib"
-projects[commons_body][version] = "3.x-dev"
 
 projects[commons_bw][type] = "module"
 projects[commons_bw][subdir] = "contrib"
-projects[commons_bw][version] = "3.x-dev"
 
 projects[commons_content_moderation][type] = "module"
 projects[commons_content_moderation][subdir] = "contrib"
-projects[commons_content_moderation][version] = "3.x-dev"
 
 projects[commons_documents][type] = "module"
 projects[commons_documents][subdir] = "contrib"
-projects[commons_documents][version] = "3.x-dev"
 
 projects[commons_groups][type] = "module"
 projects[commons_groups][subdir] = "contrib"
-projects[commons_groups][version] = "3.x-dev"
 
 projects[commons_events][type] = "module"
 projects[commons_events][subdir] = "contrib"
-projects[commons_events][version] = "3.x-dev"
 
 projects[commons_featured][type] = "module"
 projects[commons_featured][subdir] = "contrib"
-projects[commons_featured][version] = "3.x-dev"
 
 projects[commons_follow][type] = "module"
 projects[commons_follow][subdir] = "contrib"
-projects[commons_follow][version] = "3.x-dev"
 
 projects[commons_location][type] = "module"
 projects[commons_location][subdir] = "contrib"
-projects[commons_location][version] = "3.x-dev"
 
 projects[commons_misc][type] = "module"
 projects[commons_misc][subdir] = "contrib"
-projects[commons_misc][version] = "3.x-dev"
-
 
 projects[commons_notices][type] = "module"
 projects[commons_notices][subdir] = "contrib"
-projects[commons_notices][version] = "3.x-dev"
 
 projects[commons_notify][type] = "module"
 projects[commons_notify][subdir] = "contrib"
-projects[commons_notify][version] = "3.x-dev"
 
 projects[commons_pages][type] = "module"
 projects[commons_pages][subdir] = "contrib"
-projects[commons_pages][version] = "3.x-dev"
 
 projects[commons_posts][subdir] = "contrib"
 projects[commons_posts][type] = "module"
-projects[commons_posts][download][type] = "git"
-projects[commons_posts][download][url] = "http://git.drupal.org/project/commons_posts.git"
-projects[commons_posts][download][branch] = "7.x-3.x"
-
 
 projects[commons_polls][type] = "module"
 projects[commons_polls][subdir] = "contrib"
-projects[commons_polls][version] = "3.x-dev"
 
 projects[commons_profile_base][type] = "module"
 projects[commons_profile_base][subdir] = "contrib"
-projects[commons_profile_base][version] = "3.x-dev"
 
 projects[commons_profile_social][type] = "module"
 projects[commons_profile_social][subdir] = "contrib"
-projects[commons_profile_social][version] = "3.x-dev"
 
 projects[commons_q_a][type] = "module"
 projects[commons_q_a][subdir] = "contrib"
-projects[commons_q_a][version] = "3.x-dev"
 
 projects[commons_radioactivity][type] = "module"
 projects[commons_radioactivity][subdir] = "contrib"
-projects[commons_radioactivity][version] = "3.x-dev"
 
 projects[commons_like][type] = "module"
 projects[commons_like][subdir] = "contrib"
-projects[commons_like][version] = "3.x-dev"
 
 projects[commons_search][type] = "module"
 projects[commons_search][subdir] = "contrib"
-projects[commons_search][version] = "3.x-dev"
+
+projects[commons_social_sharing][type] = "module"
+projects[commons_social_sharing][subdir] = "contrib"
 
 projects[commons_site_homepage][type] = "module"
 projects[commons_site_homepage][subdir] = "contrib"
-projects[commons_site_homepage][version] = "3.x-dev"
 
 projects[commons_utility_links][type] = "module"
 projects[commons_utility_links][subdir] = "contrib"
-projects[commons_utility_links][version] = "3.x-dev"
 
 projects[commons_user_profile_pages][type] = "module"
 projects[commons_user_profile_pages][subdir] = "contrib"
-projects[commons_user_profile_pages][version] = "3.x-dev"
 
 projects[commons_wikis][type] = "module"
 projects[commons_wikis][subdir] = "contrib"
-projects[commons_wikis][version] = "3.x-dev"
 
 projects[commons_wysiwyg][type] = "module"
 projects[commons_wysiwyg][subdir] = "contrib"
 
 projects[commons_topics][type] = "module"
 projects[commons_topics][subdir] = "contrib"
-projects[commons_topics][version] = "3.x-dev"
 
 ; Contributed themes.
 
 projects[adaptivetheme][type] = "theme"
 projects[adaptivetheme][subdir] = "contrib"
-projects[adaptivetheme][version] = "3.x-dev"
+projects[adaptivetheme][download][type] = "git"
+projects[adaptivetheme][download][url] = "git://git.drupal.org/project/adaptivetheme.git"
+projects[adaptivetheme][download][branch] = "7.x-3.x"
+projects[adaptivetheme][download][revision] = "4e29d43"
 
 projects[sky][type] = "theme"
 projects[sky][subdir] = "contrib"
+projects[sky][version] = "3.0-rc1"
 
 projects[commons_origins][type] = "theme"
-projects[commons_origins][version] = "3.x-dev"
 projects[commons_origins][subdir] = "contrib"
 
 ; Libraries.
diff --git a/images/avatars/avatar-al_steffen.png b/images/avatars/avatar-al_steffen.png
new file mode 100644
index 0000000..28d1fac
Binary files /dev/null and b/images/avatars/avatar-al_steffen.png differ
diff --git a/images/avatars/avatar-dharmesh_mistry.png b/images/avatars/avatar-dharmesh_mistry.png
new file mode 100644
index 0000000..ded5899
Binary files /dev/null and b/images/avatars/avatar-dharmesh_mistry.png differ
diff --git a/images/avatars/avatar-drew_robertson.png b/images/avatars/avatar-drew_robertson.png
new file mode 100644
index 0000000..f3e3b5b
Binary files /dev/null and b/images/avatars/avatar-drew_robertson.png differ
diff --git a/images/avatars/avatar-erica_ligeski.png b/images/avatars/avatar-erica_ligeski.png
new file mode 100644
index 0000000..3d93f0f
Binary files /dev/null and b/images/avatars/avatar-erica_ligeski.png differ
diff --git a/images/avatars/avatar-james_elliott.png b/images/avatars/avatar-james_elliott.png
new file mode 100644
index 0000000..0e7c0d2
Binary files /dev/null and b/images/avatars/avatar-james_elliott.png differ
diff --git a/images/avatars/avatar-jeff_noyes.png b/images/avatars/avatar-jeff_noyes.png
new file mode 100644
index 0000000..5ade4bf
Binary files /dev/null and b/images/avatars/avatar-jeff_noyes.png differ
diff --git a/images/avatars/avatar-katelyn_fogarty.png b/images/avatars/avatar-katelyn_fogarty.png
new file mode 100644
index 0000000..f0ee03c
Binary files /dev/null and b/images/avatars/avatar-katelyn_fogarty.png differ
diff --git a/images/avatars/avatar-kevin_oleary.png b/images/avatars/avatar-kevin_oleary.png
new file mode 100644
index 0000000..f250cb1
Binary files /dev/null and b/images/avatars/avatar-kevin_oleary.png differ
diff --git a/images/avatars/avatar-lisa_rex.png b/images/avatars/avatar-lisa_rex.png
new file mode 100644
index 0000000..556f7cd
Binary files /dev/null and b/images/avatars/avatar-lisa_rex.png differ
diff --git a/images/avatars/user-avatar.png b/images/avatars/user-avatar.png
new file mode 100644
index 0000000..ffe5cd6
Binary files /dev/null and b/images/avatars/user-avatar.png differ
diff --git a/images/avatars/user-avatar.svg b/images/avatars/user-avatar.svg
new file mode 100644
index 0000000..67d4152
--- /dev/null
+++ b/images/avatars/user-avatar.svg
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 16.0.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
+	 width="145px" height="145px" viewBox="0 0 145 145" enable-background="new 0 0 145 145" xml:space="preserve">
+<g>
+	<g>
+		<rect fill="#CBCBCB" width="145" height="145"/>
+	</g>
+	<g id="layer1" transform="matrix(1.0488323,0,0,1,-5.6511307,-963.98203)">
+		<path id="path5" inkscape:connector-curvature="0" fill="#010101" d="M12.18,1048.865"/>
+		<path id="path13" inkscape:connector-curvature="0" fill="#F9F9FA" d="M43.084,1053.027
+			c-19.986,0.082-37.606,35.246-37.537,56.781c0,0,103.477,0.062,136.024,0c0.031-19.98-16.795-56.664-38.681-56.568
+			L43.084,1053.027z"/>
+		<path id="path15" inkscape:connector-curvature="0" fill="#F9F9FA" d="M73.193,1044.648c15.625,0.09,28.36-13.533,28.428-30.42
+			c-0.068-16.73-12.803-30.352-28.428-30.417c-15.786,0.065-28.522,13.687-28.419,30.413
+			C44.671,1031.115,57.406,1044.734,73.193,1044.648L73.193,1044.648L73.193,1044.648z"/>
+		<path id="path17" inkscape:connector-curvature="0" fill="#010101" d="M55.211,1043.201"/>
+	</g>
+</g>
+</svg>
diff --git a/images/commons_group_description_sample.png b/images/commons_group_description_sample.png
old mode 100644
new mode 100755
diff --git a/images/commons_homepage_text_example.png b/images/commons_homepage_text_example.png
old mode 100644
new mode 100755
